concurrency: include safe details in `discoveredLock()` assertion

erikgrinaker opened the pull request:
As seen in #63592, an assertion in `discoveredLock` occasionally fails,
but does not include enough details to debug the issue.

This patch implements `SafeFormatter` for `lockState`, and includes the
redacted lock state and txn in the assertion error message.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64130)
<!-- Reviewable:end -->

nvb reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-:-MZ-aibW1bMv-wCnvlh1:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-b7icCyIwrSbeVAKg:-MZ-b7icCyIwrSbeVAKh:b7c7wkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table.go#L1520)):*
> ```Go
> 			return errors.AssertionFailedf(
> 				"discovered lock by different transaction (%s) than existing lock: %s",
> 				errors.Safe(txn.ID), l,
> ```

Can we print more information about `txn` here? `TxnMeta` should already implement `SafeMessager`.
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-bUSW-FKHKwAxRqhl:-MZ-bUSW-FKHKwAxRqhm:bc2tn37) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table.go#L1520)):*
> ```Go
> 			return errors.AssertionFailedf(
> 				"discovered lock by different transaction (%s) than existing lock: %s",
> 				errors.Safe(txn.ID), l,
> ```

Should we make `uuid.UUID` implement `redact.SafeValue`?
___
*[pkg/kv/kvserver/concurrency/lock_table_test.go, line 1319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-abFU6mAQm8Qs2-Ee:-MZ-abFU6mAQm8Qs2-Ef:baptjpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table_test.go#L1319)):*
> ```Go
> 		seqs: []enginepb.TxnSeq{1},
> 	}
> 	require.EqualValues(t,
> ```

Should we also assert the output of `redact.Sprint(l)`, like we do above?


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
concurrency: include safe details in `discoveredLock()` assertion

As seen in #63592, an assertion in `discoveredLock` occasionally fails,
but does not include enough details to debug the issue.

This patch implements `SafeFormatter` for `lockState`, and includes the
redacted lock state and txn in the assertion error message.

Release note: None

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64130)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-b7icCyIwrSbeVAKg:-MZ32SLi5MvrVv54WoRm:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table.go#L1520)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we print more information about `txn` here? `TxnMeta` should already implement `SafeMessager`.
</blockquote></details>

Good idea, done.
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-bUSW-FKHKwAxRqhl:-MZ32VmKCLUAdH79bWB2:bdzrf9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table.go#L1520)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we make `uuid.UUID` implement `redact.SafeValue`?
</blockquote></details>

Suppose we could, but it's not necessary here since we use `TxnMeta` instead.
___
*[pkg/kv/kvserver/concurrency/lock_table_test.go, line 1319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-MZ-abFU6mAQm8Qs2-Ee:-MZ32fPE3bCC1Rdgn80E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e784447145107e45b9e00ac6817e8af791b87112/pkg/kv/kvserver/concurrency/lock_table_test.go#L1319)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we also assert the output of `redact.Sprint(l)`, like we do above?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64130#-:-MZDc6oT7Gj4wVqF19kG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
bors r=nvanbenschoten

erikgrinaker commented:
bors r=nvanbenschoten
